src/lib.rs

```diff
@@ -62,193 +106,250 @@ pub fn transcode<'de, D, S>(d: D, s: S) -> Result<S::Ok, S::Error>
 /// Unlike traditional serializable types, `Transcoder`'s `Serialize`
 /// implementation is *not* idempotent, as it advances the state of its
 /// internal `Deserializer`. It should only ever be serialized once.
-pub struct Transcoder<D>(RefCell<Option<D>>);
+pub struct Transcoder<'de, D>
+    where D: de::Deserializer<'de>
+{
+    de: RefCell<Option<D>>,
+    de_err: RefCell<Option<D::Error>>,
+}
 
-impl<'de, D> Transcoder<D>
+impl<'de, D> Transcoder<'de, D>
     where D: de::Deserializer<'de>
 {
     /// Constructs a new `Transcoder`.
-    pub fn new(d: D) -> Transcoder<D> {
-        Transcoder(RefCell::new(Some(d)))
+    pub fn new(de: D) -> Transcoder<'de, D> {
+        Transcoder {
+            de: RefCell::new(Some(de)),
+            de_err: RefCell::new(None),
+        }
+    }
+
+    /// Returns the last error that has occured on the `Deserializer` side
+    /// while transcoding. Errors on the `Serializer` side are returned by the
+    /// `Serialize` implementation of the `Transcoder`.
+    ///
+    /// # Note
+    ///
+    /// This function should only be called once - right after the `Transcoder`
+    /// has been serialized, because it removes the error stored internally.
+    pub fn take_last_deserializer_error(&self) -> Option<D::Error> {
+        self.de_err.borrow_mut().take()
     }
 }
 
-impl<'de, D> ser::Serialize for Transcoder<D>
+impl<'de, D> ser::Serialize for Transcoder<'de, D>
     where D: de::Deserializer<'de>
 {
-    fn serialize<S>(&self, s: S) -> Result<S::Ok, S::Error>
+    fn serialize<S>(&self, ser: S) -> Result<S::Ok, S::Error>
         where S: ser::Serializer
     {
-        self.0
+        let result = self
+            .de
             .borrow_mut()
             .take()
             .expect("Transcoder may only be serialized once")
-            .deserialize_any(Visitor(s))
-            .map_err(d2s)
+            .deserialize_any(Visitor(ser));
+        match result {
+            Ok(Ok(v)) => Ok(v),
+            Ok(Err(ser_err)) => Err(ser_err),
+            Err(de_err) => {
+                let ser_err = <S::Error as ser::Error>::custom(de_err.to_string());
```

Comment:
It might be better to have the ser_err explicitly point to the de_error rather than just copying its message to help identify improper usage.

Reply:
What do you mean? Calling `custom` on `de_err` rather than `de_err.to_string()`? `de_err` must be moved into `self.de_err` on the next line.